### PR TITLE
Fix overlay bug with multiple field patches

### PR DESCRIPTION
**Pull request by wu-hui**
Fix: https://github.com/firebase/firebase-android-sdk/issues/3621

### CODE REVIEW

**Comment by ehsannas** [comment]
Also need a CHANGELOG entry?

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Overlay.java [hunk 1]
```diff
@@ -36,4 +36,10 @@ public static Overlay create(int largestBatchId, Mutation mutation) {
   public DocumentKey getKey() {
     return getMutation().getKey();
   }
+
+  public FieldMask getFieldMask() {
+    return getMutation() instanceof PatchMutation
+        ? ((PatchMutation) getMutation()).getMask()
+        : null;
+  }
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/Overlay.java]: Done.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java [hunk 2]
```diff
@@ -119,29 +119,36 @@ Document getDocument(DocumentKey key) {
...
     Map<DocumentKey, Overlay> overlays = new HashMap<>();
     populateOverlays(overlays, docs.keySet());
-    return computeViews(docs, overlays, existenceStateChanged);
+    ImmutableSortedMap<DocumentKey, Document> result = emptyDocumentMap();
+    for (Map.Entry<DocumentKey, OverlayedDocument> entry :
+        computeViews(docs, overlays, existenceStateChanged).entrySet()) {
+      result = result.insert(entry.getKey(), entry.getValue().getDocument());
+    }
+
+    return result;
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java:128]: I believe sorting by key is more for representing to user in a consistent manner.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java [hunk 3]
```diff
@@ -142,7 +142,16 @@ Document getDocument(DocumentKey key) {
...
   }
 
-  /*Computes the local view for doc */
+  /**
+   * Computes the local view for the given documents.
+   *
+   * @param docs The documents to compute views for. It also has the base version of the documents.
+   * @param overlays The overlays that need to be applied to the given base version of the
+   *     documents.
+   * @param existenceStateChanged A set of documents that might there existence states changed. This
```
Comment (ehsannas) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java:151]: s/that might there existence states changed/whose existence states might have changed.

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java [hunk 4]
```diff
@@ -100,14 +100,11 @@ public void applyToRemoteDocument(MutableDocument document, MutationBatchResult
   }
 
   /**
-   * Computes the local view of a document given all the mutations in this batch. Returns an {@link
-   * FieldMask} representing all the fields that are mutated.
+   * Computes the local view of a document given all the mutations in this batch.
+   *
+   * @param mutatedFields Fields that are already mutated before applying the current one.
```
Comment (ehsannas) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java:105]: nit: missing `@param document`

File: firebase-firestore/src/main/java/com/google/firebase/firestore/remote/RemoteSerializer.java [hunk 5]
```diff
@@ -269,7 +269,7 @@ private MutableDocument decodeMissingDocument(BatchGetDocumentsResponse response
       builder.setUpdate(encodeDocument(mutation.getKey(), ((SetMutation) mutation).getValue()));
     } else if (mutation instanceof PatchMutation) {
       builder.setUpdate(encodeDocument(mutation.getKey(), ((PatchMutation) mutation).getValue()));
-      builder.setUpdateMask(encodeDocumentMask(((PatchMutation) mutation).getMask()));
+      builder.setUpdateMask(encodeDocumentMask(((PatchMutation) mutation).getFieldMask()));
```
Comment (ehsannas) [inline firebase-firestore/src/main/java/com/google/firebase/firestore/remote/RemoteSerializer.java:272]: casting to `PatchMutation` is not needed anymore.

**Review by ehsannas — APPROVED** [review]
LGTM, some nits below

File: firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java (hunk 4 above)
Comment (wu-hui) [reply firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/MutationBatch.java:105]: Fixed.

**Pull request merged**